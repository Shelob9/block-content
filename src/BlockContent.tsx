import { Block, parse } from '@wordpress/block-serialization-default-parser';
import React, { createElement, ReactNode, useMemo, FC } from 'react';
import { addAndParseBlock } from './transformNodes';

/**
 * Render block content with fallback to rendered.
 *
 * Safer option than default export
 */
export const RenderBlockContent: FC<{ raw?: string; rendered: string }> = ({
  raw,
  rendered,
}) => {
  function createMarkup() {
    return { __html: rendered };
  }
  if (!raw) {
    return <div dangerouslySetInnerHTML={createMarkup()} />;
  }
  return <BlockContent rawContent={raw} />;
};
export type PROPS = {
  rawContent: string;
};
/**
 * Render a block
 *
 * Only works with raw block content.
 */
const BlockContent = (props: PROPS) => {
  let { rawContent } = props;
  let Component = useMemo(() => {
    let blocks = parse(rawContent);
    let els: ReactNode[] = [];
    Object.values(blocks).forEach((block: Block) => {
      addAndParseBlock(block, els);
    });
    return () => createElement('div', {}, els);
  }, [rawContent]);

  return (
    <>
      <Component />
    </>
  );
};

export default BlockContent;
