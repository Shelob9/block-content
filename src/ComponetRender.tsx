import React,{ createElement, FC, Fragment } from "react";
import { NODE } from ".";

 const ComponetRender : FC<{node:NODE}> = ({node}) => {
    return createElement(
      node.name as string,
      {className: node.attribs && node.attribs.class ? node.attribs.class : undefined},
      node.children ? node.children.map(
        (child : NODE,i:number) => {
            return(
                <Fragment key={i}>
                    {'text' === child.type ? (
                        <Fragment>{child.data}</Fragment>
                    ) : (
                        <ComponetRender node={child} />
                    )}
                </Fragment>
            )
        }
      ) : []
    )
  }

  export default ComponetRender;