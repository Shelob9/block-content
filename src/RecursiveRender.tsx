import React,{ createElement, FC, Fragment } from "react";
import { NODE } from ".";

 const RecursiveRender : FC<{node:NODE}> = ({node}) => {
    return createElement(
      node.name as string,
      {className: node.attribs && node.attribs.class ? node.attribs.class : undefined},
      node.children ? node.children.map(
        (child : NODE,i:number) => {
            return(
                <Fragment key={i}>
                    {'text' === child.type ? (
                        <Fragment>{child.data}</Fragment>
                    ) : (
                        <RecursiveRender node={child} />
                    )}
                </Fragment>
            )
          
          
  
        }
      ) : []
    )
  }

  export default RecursiveRender;