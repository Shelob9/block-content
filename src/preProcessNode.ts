import { NODE } from '.';

/**
 * Prepares nodes with type "tag" to be rendered.
 *
 * Enforces:
 *  - Allowed node types
 *  - Allowed node children
 *  - Allowed attributes.
 */
export const preProcessTagNode = (node: NODE) => {
  switch (node.name) {
    case 'i':
      node.name = 'em';
      break;
    default:
      break;
  }
  return node;
};
