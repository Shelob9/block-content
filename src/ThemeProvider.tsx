import React, { createContext, FC } from 'react';
import { ComponetsMap } from './ComponetRender';

export type tagSettings = {
  [key: string]: string[];
};

export const ThemeContext = createContext<{
  components: ComponetsMap;
  tagSettings?: tagSettings;
}>(
  //@ts-ignore
  null
);

const ThemeProvider: FC<{
  children: any;
  components: ComponetsMap;
  tagSettings?: tagSettings;
}> = ({ children, components, tagSettings }) => {
  return (
    <ThemeContext.Provider value={{ components, tagSettings }}>
      {children}
    </ThemeContext.Provider>
  );
};

export default ThemeProvider;
