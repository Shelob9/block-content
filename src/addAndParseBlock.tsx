import { Block } from '@wordpress/block-serialization-default-parser';
import React, { createElement, Fragment, ReactElement, ReactNode } from 'react';
import ReactHtmlParser from 'react-html-parser';
import ComponetRender from './ComponetRender';
import { preProcessTagNode } from './preProcessNode';

export const parseHtml = (innerHTML: string) => {
  return ReactHtmlParser(innerHTML, {
    transform,
  });
};
/**
 * Node being parsed to element
 */
export type NODE = {
  type: 'style' | 'tag' | 'text' | 'script';
  name?: string;
  attribs: any;
  children?: NODE[];
  next?: NODE;
  data?: string;
};

//Transform function for react-html-parser
export function transform(node: NODE, index: number): ReactElement {
  if ('style' === node.type || 'script' === node.type) {
    return <Fragment key={index} />;
  }
  if ('tag' === node.type) {
    node = preProcessTagNode(node);
    return (
      <Fragment key={index}>
        <ComponetRender node={node} />
      </Fragment>
    );
  }
  return <Fragment key={index} />;
}

//@todo make this step not needed.
const addAndParseBlock = (block: Block, els: ReactNode[]) => {
  if (block.innerBlocks.length) {
    //Special handling for columns
    // @todo use same render function as other elements
    switch (block.blockName) {
      case 'core/columns':
        let innerBlockEls: ReactElement[] = [];
        Object.values(block.innerBlocks).forEach((innerBlock: Block) => {
          addAndParseBlock(innerBlock, innerBlockEls);
        });
        els.push(
          createElement(
            'div',
            {
              class: `wp-block-columns has-${block.attrs.columns}-columns`,
            },
            innerBlockEls
          )
        );
        return els;
      case 'core/column': {
        let innerBlockEls: ReactElement[] = [];
        Object.values(block.innerBlocks).forEach((innerBlock: Block) => {
          addAndParseBlock(innerBlock, innerBlockEls);
        });
        els.push(
          createElement(
            'div',
            {
              class: `wp-block-column`,
            },
            innerBlockEls
          )
        );
        return els;
      }
    }
  } else {
    els.push(parseHtml(block.innerHTML));
  }
  return els;
};

export default addAndParseBlock;
