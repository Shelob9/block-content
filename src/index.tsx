import BlockContent, { RenderBlockContent } from './BlockContent';

import ThemeProvider from './ThemeProvider';

export { ThemeProvider, RenderBlockContent };

export default BlockContent;
