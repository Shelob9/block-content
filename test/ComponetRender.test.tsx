import React from 'react';
import { render } from '@testing-library/react';
import ComponetRender from '../src/ComponetRender';
import { NODE } from '../src';

describe('ComponetRender', () => {
  it('renders paragraph with text children', () => {
    let node : NODE = {
      type: 'tag',
      name: 'p',
      attribs: {
        class: 'food'
      },
      children: [
        {
          type: 'text',
          data: 'Hi Roy',
          attribs: {}
        },
      ]
    }
    const { container,getByText } = render(
    <ComponetRender node={node} />);
    expect(getByText('Hi Roy')).toBeTruthy();

    expect(container.querySelectorAll('p').length).toBe(1);
    expect(container.querySelectorAll('.food').length).toBe(1);

  });
  it('renders paragraph with tag children', () => {
    let node : NODE = {
      type: 'tag',
      name: 'p',
      attribs: {
        class: 'food'
      },
      children: [
        {
          type: 'tag',
          name: 'em',
          attribs: {
            class: 'food'
          },
          children: [
            {
              type: 'text',
              data: 'Hi Roy',
              attribs: {}
            },
          ]
        }
      ]
    }
    const { container,getByText } = render(
    <ComponetRender node={node} />);
    

    expect(container.querySelectorAll('p').length).toBe(1);
    expect(container.querySelectorAll('em').length).toBe(1);

    expect(container.querySelectorAll('.food').length).toBe(2);
    expect(getByText('Hi Roy')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
  
});
