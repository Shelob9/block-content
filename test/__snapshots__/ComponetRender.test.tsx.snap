// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponetRender renders paragraph with tag children 1`] = `
<div>
  <p
    class="food"
  >
    <em
      class="food"
    >
      Hi Roy
    </em>
  </p>
</div>
`;
